File: Umix/emulator/cpu.cc
#include "emulator/cpu.h"

#include "emulator/mem.h"
#include "emulator/instruction.h"
#include "emulator/terminal.h"

CPU::CPU(Mem* mem, Terminal* terminal)
    : mem_(mem),
      terminal_(terminal) {}

bool CPU::ExecuteSingleInstruction() {
  uint32_t instruction_code = mem_->Read(0, instruction_pointer_);
  Instruction instr = Instruction::Decode(instruction_code);
  instruction_pointer_++;
  // printf("Executing instruction: %s\n", instr.ToString().c_str());
  switch (instr.op_code) {
    case Instruction::CMOV: {
        if (regs_[instr.r_c] != 0)
          regs_[instr.r_a] = regs_[instr.r_b];
      }
      break;
    case Instruction::LD: {
        regs_[instr.r_a] = mem_->Read(regs_[instr.r_b], regs_[instr.r_c]);
      }
      break;
    case Instruction::ST: {
        mem_->Write(regs_[instr.r_a], regs_[instr.r_b], regs_[instr.r_c]);
      }
      break;
    case Instruction::ADD: {
        regs_[instr.r_a] = regs_[instr.r_b] + regs_[instr.r_c];
      }
      break;
    case Instruction::MUL: {
        regs_[instr.r_a] = regs_[instr.r_b] * regs_[instr.r_c];
      }
      break;
    case Instruction::DIV: {
        regs_[instr.r_a] = regs_[instr.r_b] / regs_[instr.r_c];
      }
      break;
    case Instruction::NOT_AND: {
        regs_[instr.r_a] = ~(regs_[instr.r_b] & regs_[instr.r_c]);
      }
      break;
    case Instruction::HALT: {
        return false;
      }
      break;
    case Instruction::ALLOC: {
        regs_[instr.r_b] = mem_->AllocPage(regs_[instr.r_c]);
      }
      break;
    case Instruction::FREE: {
        mem_->FreePage(regs_[instr.r_c]);
      }
      break;
    case Instruction::OUT: {
        terminal_->Write(regs_[instr.r_c]);
      }
      break;
    case Instruction::IN: {
        regs_[instr.r_c] = terminal_->Read();
      }
      break;
    case Instruction::LOAD_PROG: {
        if (regs_[instr.r_b] != 0)
          mem_->CopyPageToZeroPage(regs_[instr.r_b]);
        instruction_pointer_ = regs_[instr.r_c];
      }
      break;
    case Instruction::IMM: {
        regs_[instr.r_a] = instr.immed;
      }
      break;
    default:
      printf("Incorrect instruction\n");
      return false;
  }
  return true;
}

File: Umix/emulator-rs/src/cpu.rs
use instruction;
use mem;
use terminal;

pub struct CPU<'a> {
  regs: [u32;8],
  instruction_pointer: u32,
  mem: &'a mut mem::Mem,
  terminal: &'a terminal::Terminal,
}

impl<'a> CPU<'a> {
  pub fn new(mem: &'a mut mem::Mem,
             terminal: &'a terminal::Terminal) -> Self {
    CPU {
      regs: [0; 8],
      instruction_pointer: 0,
      mem,
      terminal,
    }
  }

  pub fn execute_single_instruction(&mut self) -> bool {
    let instruction_code = self.mem.read(0, self.instruction_pointer);
    self.instruction_pointer += 1;
    let instr = instruction::decode(instruction_code);
    match instr.op_code {
      instruction::CMOV => {
        if self.regs[instr.r_c] != 0 {
          self.regs[instr.r_a] = self.regs[instr.r_b];
        }
      }
      instruction::LD => {
        self.regs[instr.r_a] = self.mem.read(
            self.regs[instr.r_b], self.regs[instr.r_c]);
      }
      instruction::ST => {
        self.mem.write(
            self.regs[instr.r_a], self.regs[instr.r_b], self.regs[instr.r_c]);
      }
      instruction::ADD => {
        let b: u64 = self.regs[instr.r_b] as u64;
        let c: u64 = self.regs[instr.r_c] as u64;
        self.regs[instr.r_a] = (b + c) as u32;
      }
      instruction::MUL => {
        let b: u64 = self.regs[instr.r_b] as u64;
        let c: u64 = self.regs[instr.r_c] as u64;

        self.regs[instr.r_a] = (b * c) as u32;
      }
      instruction::DIV => {
        self.regs[instr.r_a] = self.regs[instr.r_b] / self.regs[instr.r_c];
      }
      instruction::NOT_AND => {
        self.regs[instr.r_a] = !(self.regs[instr.r_b] & self.regs[instr.r_c]);
      }
      instruction::HALT => {
        return false;
      }
      instruction::ALLOC => {
        self.regs[instr.r_b] = self.mem.alloc_page(self.regs[instr.r_c]);
      }
      instruction::FREE => {
        self.mem.free_page(self.regs[instr.r_c]);
      }
      instruction::OUT => {
        self.terminal.write(self.regs[instr.r_c]);
      }
      instruction::IN => {
        self.regs[instr.r_c] = self.terminal.read();
      }
      instruction::LOAD_PROG => {
        if self.regs[instr.r_b] != 0 {
          self.mem.copy_page_to_zero_page(self.regs[instr.r_b]);
        }
        self.instruction_pointer = self.regs[instr.r_c];
      }
      instruction::IMM => {
        self.regs[instr.r_a] = instr.immed;
      }
      _ => {
        assert!(false);
      }
    }
    true
  }
}